docs/blog/2021-06-16-tailor-your-slack-audit-sink-messages.md

```diff
@@ -0,0 +1,169 @@
+
+That minimal configuration is all that’s needed to set up this feature for any Clutch workflow!
+
+<Image alt="Clutch Custom Slack Audit Message" src="https://user-images.githubusercontent.com/39421794/122236354-734b7380-ce8c-11eb-97e4-146222eda962.png" />
+
+## Under the Hood
+
+In the Slack sink, we check if a user has provided a custom Slack message for a given workflow  event and if so, inject the metadata values from the audit event into the custom message and [execute](https://golang.org/pkg/text/template/#Template.Execute) the message template.
+
+As mentioned earlier, one of our goals for the customized Slack message feature was to reduce code duplication. By saving the API request/response body as Any type in the audit event, we can retrieve the metadata values for any custom message with just [this flow](https://github.com/lyft/clutch/blob/9765a88f042c306b0a9d63fb55a8b2ed552b1c18/backend/service/auditsink/slack/slack.go#L174-L202) in the Slack sink. Let’s focus on the important pieces.
+
+```go
+reqJSON, err := protojson.Marshal(event.RequestMetadata.Body)
+respJSON, err := protojson.Marshal(event.ResponseMetadata.Body)
+```
+
+The aim here is to unpack the metadata (Any type) into its underlying message. We looked into using anypb’s [UnmarshalNew](https://pkg.go.dev/google.golang.org/protobuf/types/known/anypb#Any.UnmarshalNew) but ultimately decided upon protojson’s [Marshal](https://pkg.go.dev/google.golang.org/protobuf/encoding/protojson#Marshal). The main reason for this choice was that we wanted to use the json format as the basis for the context in order to simplify the metadata structure a bit. For example, protojson flattens `oneOf` fields into its top level field. Since custom messages need to specify what exact data is needed from the request/response, simplifications like these can help make writing a custom message easier.
+
+```go
+var requestMetadata map[string]interface{}
+var responseMetadata map[string]interface{}
+
+json.Unmarshal(reqJSON, &requestMetadata)
+json.Unmarshal(respJSON, &responseMetadata)
+```
+
+API request/response messages can be complex types, so we unmarshaled the metadata into a `map[string]interface{}`. This form can be used with many other packages and in particular we wanted to support any of the Golang template [functions](https://golang.org/pkg/text/template/#hdr-Functions) (and make it easy for contributors to enable template helpers like [Sprig](http://masterminds.github.io/sprig/)) so that users can easily retrieve the metadata for their custom Slack audits.
+
+```go
+return &auditTemplateData{
+	Request:  requestMetadata,
+	Response: responseMetadata,
+
+}
+```
+
+Lastly, we didn’t want the users to need to know anything about the schema of the audit event metadata, so we saved the metadata in a new simple struct. In this way, the user can chain the information they wanted to retrieve with just `.Request.<field name>` or `.Response.<field name>`.
```

Comment:
Would it help to add a snippet of what the "new simple struct" is?

Reply:
the snippet right above it is it. 

```go
return &auditTemplateData{
	Request:  requestMetadata,
	Response: responseMetadata,

}
```
for each code block, i follow with an explanation. lmk if that pattern is confusing/if you have thoughts on how to make it more clear